unison-cli/src/Unison/Codebase/Editor/HandleInput/AuthLogin.hs

```diff
@@ -1,28 +1,218 @@
+    ("GET", ["redirect"], (Just code, maybeNextURI)) -> do
+      callback code maybeNextURI >>= respond
+    _ -> respond (responseLBS status404 [] "Not Found")
+  where
+    getQueryParams req = do
+      let code = join $ Prelude.lookup "code" (queryString req)
+          nextURI = do
+            nextBS <- join $ Prelude.lookup "next" (queryString req)
+            parseURI (BSC.unpack nextBS)
+       in (Text.decodeUtf8 <$> code, nextURI)
+
+-- | Construct an authorization URL from the parameters required.
+authURI :: URI -> String -> OAuthState -> PKCEChallenge -> URI
+authURI authEndpoint redirectURI state challenge =
+  authEndpoint
+    & addQueryParam "state" state
+    & addQueryParam "redirect_uri" (BSC.pack redirectURI)
+    & addQueryParam "response_type" "code"
+    & addQueryParam "scope" "openid cloud sync"
+    & addQueryParam "client_id" ucmOAuthClientID
+    & addQueryParam "code_challenge" challenge
+    & addQueryParam "code_challenge_method" "S256"
+
+addQueryParam :: ByteString -> ByteString -> URI -> URI
+addQueryParam key val uri =
+  let existingQuery = parseQuery $ BSC.pack (uriQuery uri)
+      newParam = (key, Just val)
+   in uri {uriQuery = BSC.unpack $ renderQuery True (existingQuery <> [newParam])}
+
+generateParams :: MonadIO m => m (PKCEVerifier, PKCEChallenge, OAuthState)
+generateParams = liftIO $ do
+  verifier <- BE.convertToBase @ByteString BE.Base64URLUnpadded <$> getRandomBytes 50
+  let digest = Crypto.hashWith Crypto.SHA256 verifier
+  let challenge = BE.convertToBase BE.Base64URLUnpadded digest
+  state <- BE.convertToBase @ByteString BE.Base64URLUnpadded <$> getRandomBytes 12
+  pure (verifier, challenge, state)
+
+-- | Exchange an authorization code for tokens.
+exchangeCode ::
+  MonadIO m =>
+  HTTP.Manager ->
+  URI ->
+  Code ->
+  PKCEVerifier ->
+  String ->
+  m (Either CredentialFailure Tokens)
+exchangeCode httpClient tokenEndpoint code verifier redirectURI = liftIO $ do
+  req <- HTTP.requestFromURI tokenEndpoint
+  let addFormData =
+        urlEncodedBody
+          [ ("code", Text.encodeUtf8 code),
+            ("code_verifier", verifier),
+            ("grant_type", "authorization_code"),
+            ("redirect_uri", BSC.pack redirectURI),
+            ("client_id", ucmOAuthClientID)
+          ]
+  let fullReq = addFormData $ req {HTTP.method = "POST", HTTP.requestHeaders = [("Accept", "application/json")]}
+  resp <- HTTP.httpLbs fullReq httpClient
+  case HTTP.responseStatus resp of
+    status
+      | status < status300 -> do
+        let respBytes = HTTP.responseBody resp
+        pure $ case Aeson.eitherDecode @Tokens respBytes of
+          Left err -> Left (InvalidTokenResponse tokenEndpoint (Text.pack err))
+          Right a -> Right a
+      | otherwise -> pure $ Left (InvalidTokenResponse tokenEndpoint $ "Received " <> tShow status <> " response from token endpoint")
+
+-- authenticateCodeserverCli :: CodeserverURI -> Cli r ()
```

Comment:
What's going on here? I assumed this was just a transplanted comment, but this string doesn't show up on trunk at all 🤔